src/nvme/mi.c

```diff
@@ -580,6 +603,36 @@ static int nvme_mi_admin_parse_status(struct nvme_mi_resp *resp, __u32 *result)
 	return nvme_status;
 }
 
+static int nvme_mi_service_parse_status(struct nvme_mi_resp *resp, __le16 *cpsr)
+{
+	struct nvme_mi_service_resp_hdr *resp_hdr;
+
+	/* we have a few different sources of "result" here: the status header
+	 * in the MI response, the cdw3 status field, and (command specific)
+	 * return values in cdw0. The latter is returned in the result pointer,
+	 * the former two generate return values here
+	 */
```

Comment:
This comment looks to have been copy-pasted from elsewhere, but doesn't apply to the control primitive messages: there is no cdw0/cdw3 status.

Reply:
>  I'm not clear on what you mean by "breaked" command, could you elaborate on that?

yep, see NVM-Express-Management-Interface-Specification-1.2d-2024.01.10-Ratified.pdf, 
`Figure 34: Command Servicing State Diagram`

Here is the English translation:

For example, when NVMe transmits messages to the Management Controller (like BMC, hereinafter referred to as BMC), the state changes to "Transmit". It remains in this state until the message transmission completes, after which it reverts to "Idle".

In cases where multiple messages are needed to assemble a complete MI message, such as the admin Identify message with an MTU of 64, receiving 4096 bytes results in approximately 60-70 packets. If the BMC stops receiving halfway through, NVMe remains in the "Transmit" state without transitioning to "Idle", preventing further communication. Consultation with NVMe vendors such as Samsung and others suggests sending an abort command to return the NVMe controller to the `Idle` state.

Regarding the scenario where BMC stops receiving halfway:
1. BMC Reboot.
2. BMC uses commands like Ctrl-C during the read process to actively abort.

Reply:
Overall, the "Transmit" mentioned above is just one possible case. There are many other scenarios that could cause NVMe to enter a paused state. For details, refer to Section 4.2.1.1 "Pause".

Reply:
> yep, see NVM-Express-Management-Interface-Specification-1.2d-2024.01.10-Ratified.pdf, `Figure 34: Command Servicing State Diagram`

OK, cool. I'd suggest referring to the states & transitions in your commit message then.

> In cases where multiple messages are needed to assemble a complete MI message, such as the admin Identify message with an MTU of 64, receiving 4096 bytes results in approximately 60-70 packets. If the BMC stops receiving halfway through, NVMe remains in the "Transmit" state without transitioning to "Idle", preventing further communication. Consultation with NVMe vendors such as Samsung and others suggests sending an abort command to return the NVMe controller to the `Idle` state.

In that example, libnvme should probably be responsible for detecting the loss of communication with the device and issuing the abort. We'll still need the patches you have here for the Control Primitive implementations, but we can't just hand the connectivity problem off to the caller.

@amboar may have some good ideas on where this recovery process may be best implemented.

> Regarding the scenario where BMC stops receiving halfway:
> 
>     1. BMC Reboot.
> 
>     2. BMC uses commands like Ctrl-C during the read process to actively abort.

Agree on (1), but not on (2): the BMC's i2c layer will still ACK all of the incoming MCTP packets.

Reply:
>  but not on (2): the BMC's i2c layer will still ACK all of the incoming MCTP packets.

Maybe because of my hardware design, the NVMe under the I2C switch being preempted by other threads causes NVMe to fail to send messages to BMC fully.

Reply:
Yes, that's a good point: if the flow is destroyed, we will release the lock on the multiplexer, which may then switch away from the NVMe endpoint, so the i2c transfer may not be acked.

Reply:
> In that example, libnvme should probably be responsible for detecting the loss of communication with the device and issuing the abort. We'll still need the patches you have here for the Control Primitive implementations, but we can't just hand the connectivity problem off to the caller.

 my current implementation involves checking the `paused` flag both when the caller fails and during initialization. Perhaps my commit was a bit rough; my main intention was to raise this issue and gather opinions, particularly from the `mi Servicing` team.

In my view, for a CLI, it's feasible to check every time. However, for a BMC, which may be constantly monitoring, I'm considering implementing some kind of error handler logic on my end.